zilliqa/src/api/eth.rs

```diff
@@ -537,38 +560,143 @@ fn send_raw_transaction(params: Params, node: &Arc<Mutex<Node>>) -> Result<Strin
         .ok_or_else(|| anyhow!("no 0x prefix"))?;
     let transaction = hex::decode(transaction)?;
     let chain_id = node.lock().unwrap().config.eth_chain_id;
-    let transaction = transaction_from_rlp(&transaction, chain_id)?;
+    let transaction = parse_transaction(&transaction)?;
+
+    if let Some(c) = transaction.chain_id() {
+        if c != chain_id {
+            return Err(anyhow!("invalid chain ID, expected: {chain_id}, got: {c}"));
+        }
+    }
 
     let transaction_hash = H256(node.lock().unwrap().create_transaction(transaction)?.0);
 
     Ok(transaction_hash.to_hex())
 }
 
-/// Decode a transaction from its RLP-encoded form.
-fn transaction_from_rlp(bytes: &[u8], chain_id: u64) -> Result<SignedTransaction> {
-    let rlp = Rlp::new(bytes);
-    let nonce = rlp.val_at(0)?;
-    let gas_price = rlp.val_at(1)?;
-    let gas_limit = rlp.val_at(2)?;
-    let to_addr = rlp.val_at::<Vec<u8>>(3)?;
-    let amount = rlp.val_at(4)?;
-    let payload = rlp.val_at(5)?;
-    let v = rlp.val_at::<u64>(6)?;
-    let r = left_pad_arr(&rlp.val_at::<Vec<_>>(7)?)?;
-    let s = left_pad_arr(&rlp.val_at::<Vec<_>>(8)?)?;
-
-    let signing_info = SigningInfo::Eth { v, r, s, chain_id };
-
-    let transaction = Transaction {
-        nonce,
-        gas_price,
-        gas_limit,
-        to_addr: (!to_addr.is_empty()).then_some(Address::from_slice(&to_addr)),
-        amount,
-        payload,
-    };
-
-    SignedTransaction::new(transaction, signing_info)
+fn parse_transaction(bytes: &[u8]) -> Result<SignedTransaction> {
+    // https://eips.ethereum.org/EIPS/eip-2718#backwards-compatibility
+    // "Clients can differentiate between the legacy transactions and typed transactions by looking at the first byte.
+    // If it starts with a value in the range [0, 0x7f] then it is a new transaction type, if it starts with a value in
+    // the range [0xc0, 0xfe] then it is a legacy transaction type."
+    match bytes[0] {
+        0xc0..=0xfe => {
```

Comment:
seems like each of these could be split into its own function for readability?

Reply:
Good idea, done